pkg/algorithm/nodeselect/v1alpha2/select_node.go

```diff
@@ -0,0 +1,193 @@
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get node list from the node selector")
+	}
+	if len(nodeList.Items) != 1 {
+		return "", errors.Errorf("could not get a unique node from the given node selectors")
+	}
+	return nodeList.Items[0].Name, nil
+}
+
+// getLabelSelectorString returns a string of label selector form label map to be used in
+// list options.
+// TODO : Move it to node package
+func getLabelSelectorString(selector map[string]string) string {
+	var selectorString string
+	for key, value := range selector {
+		selectorString = selectorString + key + "=" + value + ","
+	}
+	selectorString = selectorString[:len(selectorString)-len(",")]
+	return selectorString
+}
+
+// GetUsedNode returns a map of node for which pool has already been created.
+// Note : Filter function is not used from node builder package as it needs
+// CSP builder package which cam cause import loops.
+func (ac *Config) GetUsedNode() (map[string]bool, error) {
+	usedNode := make(map[string]bool)
+	cspList, err := csp.
+		NewKubeClient().
+		WithNamespace(ac.Namespace).
+		List(
+			metav1.
+				ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + ac.CSPC.Name},
+		)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNode[cspObj.Labels[string(apis.HostNameCPK)]] = true
+	}
+	return usedNode, nil
+}
+
+// GetBDListForNode returns a list of BD from the pool spec.
+// TODO : Move it to CStorPoolCluster packgage
+func (ac *Config) GetBDListForNode(pool *apis.PoolSpec) []string {
+	var BDList []string
+	for _, group := range pool.RaidGroups {
+		for _, bd := range group.BlockDevices {
+			BDList = append(BDList, bd.BlockDeviceName)
+		}
+	}
+	return BDList
+}
+
+// ClaimBDsForNode claims a given BlockDevice for node
+// If the block device(s) is/are already claimed for any other CSPC it returns error.
+// If the block device(s) is/are already calimed for the same CSPC -- it is left as it is and can be used for
+// pool provisioning.
+// If the block device(s) is/are unclaimed, then those are claimed.
+func (ac *Config) ClaimBDsForNode(BD []string) error {
+	for _, bdName := range BD {
+		bdAPIObj, err := bd.NewKubeClient().WithNamespace(ac.Namespace).Get(bdName, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "error in getting details for BD {%s} whether it is claimed", bdName)
+		}
+		if bd.BuilderForAPIObject(bdAPIObj).BlockDevice.IsClaimed() {
+			IsClaimedBDUsable, errBD := ac.IsClaimedBDUsable(bdAPIObj)
+			if errBD != nil {
+				return errors.Wrapf(err, "error in getting details for BD {%s} for usability", bdName)
+			}
+			if !IsClaimedBDUsable {
+				return errors.Errorf("BD {%s} already in use", bdName)
+			} else {
+				continue
+			}
+		}
+
+		err = ac.ClaimBD(bdAPIObj)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to claim BD {%s}", bdName)
+		}
+	}
+	return nil
```

Comment:
missed this point at very first time when `cspc` consider a case where there is no error in creating `bdc` but it is not claimed for some reason how this is handled?

Reply:
its handled on line#159 in claimBD() @mittachaitu .. is that what you mean?

Reply:
No It is different we will hit line #159 case at the time of reconciliation.

Reply:
Thanks @mittachaitu . Good catch.
Here is the PR
https://github.com/openebs/maya/pull/1339